content/doc/developer/extensibility/serialization-of-anonymous-classes.adoc

```diff
@@ -0,0 +1,124 @@
+
+[source]
+----
+Attempt to (de-)serialize anonymous class org.jenkinsci.plugins.stuff.YourClass$3
+in file:/…/plugins/stuff/WEB-INF/lib/stuff.jar;
+see: https://jenkins.io/redirect/serialization-of-anonymous-classes/
+----
+
+(There are variants, less common, for attempts to serialize lambdas or local classes.) The warning will be logged by `+org.jenkinsci.remoting.util.AnonymousClassWarnings+`. If you see this warning, track down the location in source code:
+
+[source]
+----
+javap -classpath /…/plugins/stuff/WEB-INF/lib/stuff.jar -l 'org.jenkinsci.plugins.stuff.YourClass$3'
+----
+
+will give you an overview of the class, and `+LineNumberTable+` entries will show you which lines in `+YourClass.java+` contain the class declaration.
+
+## Usages in Remoting
+
+The main problem with sending anonymous classes over Remoting is that can all too easily “capture” unexpected and unwanted fields in the callable, which may cause serious issues including those described in (link:https://wiki.jenkins.io/display/JENKINS/Plugins+affected+by+fix+for+JEP-200[Plugins affected by fix for JEP-200]). Also you will get unusable listings from metrics collected as of (link:https://issues.jenkins-ci.org/browse/JENKINS-27035[Request/response statistics]).
+
+To fix code like this:
+
+[source,java]
+----
+int stuff(FilePath workspace, final String arg) throws IOException, InterruptedException {
+    return workspace.act(new MasterToSlaveFileCallable<Integer>() {
+        @Override
+        Integer invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
+            return countThings(f, arg);
+        }
+    });
+}
+----
+
+just use a "convert anonymous to member" refactoring in your IDE, or manually:
+
+[source,java]
+----
+private static final class ThingCounter extends MasterToSlaveFileCallable<Integer> {
+    private final String arg;
+    ThingCounter(String arg) {
+        this.arg = arg;
+    }
+    @Override
+    Integer invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
+        return countThings(f, arg);
+    }
+}
+int stuff(FilePath workspace, String arg) throws IOException, InterruptedException {
+    return workspace.act(new ThingCounter(arg));
+}
+----
+
+## Usages in XStream
+
+If you have accidentally used an anonymous inner class as a field in some object serialized via XStream, your XML files are going to be a mess. This is a worse problem than for Remoting since not only can you suffer from link:../../../../blog/2018/01/13/jep-200[JEP-200-related] security blocks, but the mistake has been persisted to disk so you need to take into account possible saved settings.
```

Comment:
Can make the like absolute within the site so that it is easier to read and manage?
```suggestion
If you have accidentally used an anonymous inner class as a field in some object serialized via XStream, your XML files are going to be a mess. This is a worse problem than for Remoting since not only can you suffer from link:/blog/2018/01/13/jep-200[JEP-200-related] security blocks, but the mistake has been persisted to disk so you need to take into account possible saved settings.
```